### PR TITLE
Input orthogonal wcoeffs rather than running `orthogonalise` on input

### PR DESCRIPTION
**Pull request by mscroggs**
For many elements, it's not hard to directly make wcoeffs orthonormal. In these cases, running `orthogonalise` on an already orthonormal matrix is a waste of time.

For other elements, wcoeffs can be easily orthogonal except for a few more difficult rows. In these cases, orthogonalise can just be called on these few rows.

This PR removes the call to `orthogonalise` during initialisation of a finite element, and instead makes each element pass in an orthonormal wcoeffs. `orthogonalise` is still called on the inputs to a custom element.

### CODE REVIEW

File: cpp/basix/e-nce-rtc.cpp [hunk 1]
```diff
@@ -63,62 +63,32 @@ FiniteElement<T> basix::element::create_rtc(cell::type celltype, int degree,
   int dof = 0;
   if (tdim == 2)
   {
-    for (std::size_t d = 0; d < tdim; ++d)
-      for (int i = 0; i < ns_interval; ++i)
-        for (int j = 0; j < ns_interval; ++j)
-          wcoeffs(dof++, psize * d + i * nv_interval + j) = 1;
+    for (int i = 0; i < ns_interval; ++i)
+      for (int j = 0; j < nv_interval; ++j)
```
Comment (garth-wells) [inline cpp/basix/e-nce-rtc.cpp:67]: Add a curly bracket for readability.

File: cpp/basix/e-nce-rtc.cpp [hunk 2]
```diff
@@ -63,62 +63,32 @@ FiniteElement<T> basix::element::create_rtc(cell::type celltype, int degree,
...
-    for (std::size_t d = 0; d < tdim; ++d)
-    {
-      int n = 0;
-      std::vector<T> integrand(pts.extent(0));
-      for (std::size_t j = 0; j < integrand.size(); ++j)
-        integrand[j] = std::pow(pts(j, d), degree);
-      for (std::size_t c = 0; c < tdim; ++c)
-      {
-        if (c != d)
+    for (int i = 0; i < ns_interval; ++i)
```
Comment (garth-wells) [inline cpp/basix/e-nce-rtc.cpp:75]: Add a curly brackets for readability.

File: cpp/basix/e-nce-rtc.cpp [hunk 3]
```diff
@@ -63,62 +63,32 @@ FiniteElement<T> basix::element::create_rtc(cell::type celltype, int degree,
...
+      for (int j = 0; j < ns_interval; ++j)
+        for (int k = 0; k < nv_interval; ++k)
         {
-          for (std::size_t j = 0; j < integrand.size(); ++j)
-            integrand[j] *= std::pow(pts(j, c), indices[n]);
-          ++n;
+          wcoeffs(dof++, k * nv_interval * nv_interval + j * nv_interval + i)
```
Comment (garth-wells) [inline cpp/basix/e-nce-rtc.cpp:79]: Maybe introduce some temporaries or shorten `nv_interval` to improve readability?

File: cpp/basix/e-nce-rtc.cpp [hunk 4]
```diff
@@ -231,103 +201,36 @@ FiniteElement<T> basix::element::create_nce(cell::type celltype, int degree,
...
-  std::vector<T> integrand(pts.extent(0));
-  switch (tdim)
-  {
-  case 2:
-  {
-    for (int i = 0; i < degree; ++i)
-    {
-      for (std::size_t d = 0; d < tdim; ++d)
+    for (int i = 0; i < ns_interval; ++i)
+      for (int j = 0; j < nv_interval; ++j)
```
Comment (garth-wells) [inline cpp/basix/e-nce-rtc.cpp:209]: Add brace.

File: cpp/basix/e-serendipity.cpp [hunk 5]
```diff
@@ -141,38 +78,20 @@ impl::mdarray_t<T, 2> make_serendipity_space_3d(int degree)
   // Create coefficients for order (degree) polynomials
   impl::mdarray_t<T, 2> wcoeffs(ndofs, psize);
   int row_n = 0;
+
   for (int i = 0; i <= degree; ++i)
-    for (int j = 0; j <= degree - i; ++j)
-      for (int k = 0; k <= degree - i - j; ++k)
+    for (int j = 0; j <= degree - i + (i == 1 || i == degree ? 1 : 0); ++j)
+      for (int k = 0; k <= degree - i - j + (i == 1 || i == degree ? 1 : 0)
```
Comment (garth-wells) [inline cpp/basix/e-serendipity.cpp:84]: Add brace.

File: cpp/basix/e-serendipity.cpp [hunk 6]
```diff
@@ -239,12 +163,18 @@ impl::mdarray_t<T, 2> make_serendipity_div_space_2d(int degree)
...
             integrand[j] *= pts(j, a);
 
-        wcoeffs(2 * nv + a, psize * d + k)
+        wcoeffs(2 * nv + a, psize * d + nonzero[k])
             = std::reduce(integrand.begin(), integrand.end(), 0.0);
       }
     }
   }
 
+  basix::math::orthogonalise(
```
Comment (garth-wells) [inline cpp/basix/e-serendipity.cpp:172]: Maybe add `using namespace basix` at the top of the file?

File: cpp/basix/e-serendipity.cpp [hunk 7]
```diff
@@ -461,12 +416,18 @@ impl::mdarray_t<T, 2> make_serendipity_curl_space_2d(int degree)
...
+      impl::mdspan_t<T, 2>(
+          wcoeffs.data(),
+          std::array<std::size_t, 2>({wcoeffs.extent(0), wcoeffs.extent(1)})),
```
Comment (garth-wells) [inline cpp/basix/e-serendipity.cpp:428]: Does `std::array{wcoeffs.extent(0), wcoeffs.extent(1)}` work here?

**Review by garth-wells — COMMENTED** [review]
Looks good. Some style and readability suggestions.

File: cpp/basix/e-serendipity.cpp (hunk 7 above)
Comment (mscroggs) [reply cpp/basix/e-serendipity.cpp]: Yes

Comment (garth-wells) [reply cpp/basix/e-serendipity.cpp]: > Yes

Or maybe even `{wcoeffs.extent(0), wcoeffs.extent(1)}` is enough.

Comment (mscroggs) [reply cpp/basix/e-serendipity.cpp]: fails to compile without the std::array

Comment (garth-wells) [reply cpp/basix/e-serendipity.cpp]: I think
```cpp
impl::mdspan_t<T, 2>(wcoeffs.data(), wcoeffs.extent(0), wcoeffs.extent(1));
```
will work.

**Pull request merged**